bundles/org.openhab.binding.dlinksmarthome/src/main/java/org/openhab/binding/dlinksmarthome/internal/DLinkHNAP.java

```diff
@@ -0,0 +1,401 @@
+import java.security.NoSuchAlgorithmException;
+import java.time.Instant;
+import java.util.Formatter;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import javax.measure.Unit;
+import javax.naming.CommunicationException;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.soap.MessageFactory;
+import javax.xml.soap.SOAPElement;
+import javax.xml.soap.SOAPException;
+import javax.xml.soap.SOAPFactory;
+import javax.xml.soap.SOAPMessage;
+
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.api.ContentResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.util.BytesContentProvider;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+/**
+ * The {@link DLinkHNAP} class encapsulates the HNAP implementation for DLink thing handlers.
+ *
+ * @author Pascal Bies - Initial contribution
+ */
+public abstract class DLinkHNAP {
+
+    protected abstract void poll();
+
+    protected abstract void handleCommandAuthenticated(final ChannelUID channelUID, final Command command);
+
+    protected void updateStatus(ThingStatus status) {
+        updateStatus(status, ThingStatusDetail.NONE);
+    }
+
+    protected void updateStatus(ThingStatus status, ThingStatusDetail detail) {
+        updateStatus(status, detail, null);
+    }
+
+    protected abstract void updateStatus(ThingStatus status, ThingStatusDetail detail, final String message);
+
+    private Logger logger = LoggerFactory.getLogger(DLinkHNAP.class);
+    private String pin;
+    private String url;
+    private static final String USER = "admin";
+    private final HttpClient httpClient;
+    private ScheduledFuture<?> pollFuture;
+    private boolean invalidConfiguration = false;
+    Authentication authentication;
+    MessageFactory soapMessageFactory;
+    SOAPFactory soapFactory;
+    private static final int DETECT_POLL_S = 1;
+
+    private final Runnable poller = new Runnable() {
+        @Override
+        public void run() {
+            if (authenticate()) {
+                updateStatus(ThingStatus.ONLINE);
+                poll();
+            }
+        }
+    };
```

Comment:
This can be a method, like:
```suggestion
    private void poller() {
        if (authenticate()) {
            updateStatus(ThingStatus.ONLINE);
            poll();
        }
    }
```

And then you can pass it to the schedule as `this::poller`.

Reply:
nice!